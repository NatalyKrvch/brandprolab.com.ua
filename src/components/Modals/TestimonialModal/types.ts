import { NormalizedTestimonial } from '@/modules/TestimonialsSection';

export interface TestimonialModalProps {
  isOpen: boolean;
  onClose: () => void;
  testimonial: NormalizedTestimonial | null;
}
