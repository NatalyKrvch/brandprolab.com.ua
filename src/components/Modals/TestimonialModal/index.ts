export { default as TestimonialModal } from './TestimonialModal';
export * from './types';
