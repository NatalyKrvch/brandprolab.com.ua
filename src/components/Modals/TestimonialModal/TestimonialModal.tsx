'use client';

import Image from 'next/image';

import { ExternalLink } from '@/components/Links';
import { ModalWrapper } from '@/components/Modals/ModalWrapper';
import {
  TESTIMONIAL_MODAL_IMAGE_TEST_ID,
  TESTIMONIAL_MODAL_NAME_TEST_ID,
  TESTIMONIAL_MODAL_POSITION_TEST_ID,
  TESTIMONIAL_MODAL_TEXT_TEST_ID,
} from '@/lib/testIDs';
import { formatTextWithListItems } from '@/utils';

import type { TestimonialModalProps } from './types';

const TestimonialModal = ({
  isOpen,
  onClose,
  testimonial,
}: TestimonialModalProps) => {
  const formattedText = formatTextWithListItems(testimonial?.text || '');

  return (
    <ModalWrapper isOpen={isOpen} onClose={onClose}>
      {testimonial && (
        <div className="flex flex-col gap-2 desktop:gap-4">
          <Image
            data-testid={TESTIMONIAL_MODAL_IMAGE_TEST_ID}
            src={testimonial.clientPhotoUrl}
            alt={testimonial.clientName}
            width={100}
            height={100}
            priority
            className="absolute left-4 top-[-10px] h-96 w-96 rounded-full object-cover tablet:left-7 tablet:top-[-15px] tablet:h-108 tablet:w-108 desktop:left-30 desktop:top-[-20px] desktop:h-144 desktop:w-144"
          />

          <div className="absolute left-120 top-4 max-w-144 leading-none tablet:left-150 tablet:top-5 tablet:max-w-250 desktop:left-200 desktop:top-7 desktop:max-w-216">
            <ExternalLink href={testimonial.clientLink}>
              <p
                data-testid={TESTIMONIAL_MODAL_NAME_TEST_ID}
                className="transition-color-fast mb-2 text-lg font-semibold leading-none hover:text-teal-dark tablet:mb-1 tablet:text-xl desktop:mb-3 desktop:text-22"
              >
                {testimonial.clientName}
              </p>
            </ExternalLink>

            <p
              data-testid={TESTIMONIAL_MODAL_POSITION_TEST_ID}
              className="text-sm leading-none tablet:text-base tablet:leading-none desktop:text-lg desktop:leading-none"
            >
              {testimonial.clientPosition}
            </p>
          </div>

          <p
            data-testid={TESTIMONIAL_MODAL_TEXT_TEST_ID}
            className="mt-12 whitespace-pre-line text-sm text-gray-dark tablet:mt-14 tablet:text-base desktop:mt-80 desktop:text-lg"
          >
            {formattedText}
          </p>
        </div>
      )}
    </ModalWrapper>
  );
};

export default TestimonialModal;
