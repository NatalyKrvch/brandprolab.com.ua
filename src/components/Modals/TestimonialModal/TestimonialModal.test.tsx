import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import {
  MODAL_WRAPPER_TEST_ID,
  TESTIMONIAL_MODAL_IMAGE_TEST_ID,
  TESTIMONIAL_MODAL_NAME_TEST_ID,
  TESTIMONIAL_MODAL_POSITION_TEST_ID,
  TESTIMONIAL_MODAL_TEXT_TEST_ID,
} from '@/lib/testIDs';

import TestimonialModal from './TestimonialModal';

jest.mock('@/sanity/lib/serverClient', () => ({
  client: { fetch: jest.fn() },
}));

jest.mock('@/sanity/lib/image', () => ({
  urlForImage: (image: any) =>
    image?.asset?._ref ? `/mocked-url-${image.asset._ref}` : '/mocked-fallback',
}));

beforeAll(() => {
  const modalRoot = document.createElement('div');
  modalRoot.setAttribute('id', 'modal-root');
  document.body.appendChild(modalRoot);
});

describe('TestimonialModal', () => {
  const testimonial = {
    key: 'test-1',
    clientName: 'Jane Doe',
    clientPosition: 'Marketing Director at Company',
    clientPhotoUrl: '/test-image.jpg',
    clientLink: 'https://example.com',
    text: 'This is a testimonial.\nWith multiple lines.',
  };

  const onClose = jest.fn();

  const renderModal = (isOpen = true) =>
    render(
      <TestimonialModal
        isOpen={isOpen}
        onClose={onClose}
        testimonial={testimonial}
      />,
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders all testimonial info when open', () => {
    const { asFragment } = renderModal();

    expect(screen.getByTestId(MODAL_WRAPPER_TEST_ID)).toBeInTheDocument();
    expect(
      screen.getByTestId(TESTIMONIAL_MODAL_IMAGE_TEST_ID),
    ).toBeInTheDocument();
    expect(
      screen.getByTestId(TESTIMONIAL_MODAL_NAME_TEST_ID),
    ).toHaveTextContent(testimonial.clientName);
    expect(
      screen.getByTestId(TESTIMONIAL_MODAL_POSITION_TEST_ID),
    ).toHaveTextContent(testimonial.clientPosition);
    expect(
      screen.getByTestId(TESTIMONIAL_MODAL_TEXT_TEST_ID),
    ).toHaveTextContent('This is a testimonial.');

    expect(asFragment()).toMatchSnapshot();
  });

  it('does not render anything when closed', () => {
    const { asFragment, queryByTestId } = renderModal(false);

    expect(queryByTestId(MODAL_WRAPPER_TEST_ID)).not.toBeInTheDocument();
    expect(
      queryByTestId(TESTIMONIAL_MODAL_NAME_TEST_ID),
    ).not.toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
