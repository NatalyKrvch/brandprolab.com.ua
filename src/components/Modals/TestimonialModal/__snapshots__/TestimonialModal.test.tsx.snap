// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestimonialModal does not render anything when closed 1`] = `<DocumentFragment />`;

exports[`TestimonialModal renders all testimonial info when open 1`] = `<DocumentFragment />`;
